Media element never populates its UA shadow if it was initially created in a document without browsing context

dcrousso opened the pull request:
#### f331cc98c1de65a0a2523c674859e01c1011fc8d
<pre>
Media element never populates its UA shadow if it was initially created in a document without browsing context
<a href="https://bugs.webkit.org/show_bug.cgi?id=222657">https://bugs.webkit.org/show_bug.cgi?id=222657</a>
&lt;rdar://problem/75266631&gt;

Reviewed by Eric Carlson.

Previously, `HTMLMediaElement` relied on `didAddUserAgentShadowRoot` to call the JS `createControls`
that actually populates the UA shadow root. This did not work with `adoptNode` because in that case
the UA shadow root already existed and was already attached to the `&lt;video&gt;`, so it was not called.

Rather than rely on this single path to set up everything, instead have a `ControlsState` enum that
`HTMLMediaElement` can internally reason about to decide whether or not it needs to inject the JS
that creates `createControls` and then call it. This also has the added benefit of having those two
steps now be in the same place, whereas before they were split across two unrelated functions.

* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::HTMLMediaElement):
(WebCore::HTMLMediaElement::updateCaptionContainer):
(WebCore::HTMLMediaElement::virtualHasPendingActivity const):
(WebCore::HTMLMediaElement::configureTextTrackDisplay):
(WebCore::HTMLMediaElement::updateTextTrackDisplay):
(WebCore::HTMLMediaElement::updateTextTrackRepresentationImageIfNeeded):
(WebCore::HTMLMediaElement::configureMediaControls):
(WebCore::HTMLMediaElement::ensureMediaControls): Added.
(WebCore::HTMLMediaElement::getCurrentMediaControlsStatus):
(WebCore::HTMLMediaElement::ensureMediaControlsShadowRoot): Deleted.
(WebCore::HTMLMediaElement::ensureMediaControlsInjectedScript): Deleted.
(WebCore::HTMLMediaElement::didAddUserAgentShadowRoot): Deleted.

* LayoutTests/media/audio-controls-adoptNode.html: Added.
* LayoutTests/media/audio-controls-adoptNode-expected.txt: Added.
* LayoutTests/media/video-controls-adoptNode.html: Added.
* LayoutTests/media/video-controls-adoptNode-expected.txt: Added.

Canonical link: <a href="https://commits.webkit.org/253225@main">https://commits.webkit.org/253225@main</a>
</pre>

rniwa commented on Source/WebCore/html/HTMLMediaElement.h:
```diff
@@ -1153,6 +1150,9 @@ class HTMLMediaElement
     bool m_shouldVideoPlaybackRequireUserGesture : 1;
     bool m_volumeLocked : 1;
 
+    enum class ControlsState { None, Initializing, Ready };
```
Specify the width of `uint8_t`?

eric-carlson reviewed (approved):
r=me

dcrousso committed:
Media element never populates its UA shadow if it was initially created in a document without browsing context
https://bugs.webkit.org/show_bug.cgi?id=222657
<rdar://problem/75266631>

Reviewed by Eric Carlson.

Previously, `HTMLMediaElement` relied on `didAddUserAgentShadowRoot` to call the JS `createControls`
that actually populates the UA shadow root. This did not work with `adoptNode` because in that case
the UA shadow root already existed and was already attached to the `<video>`, so it was not called.

Rather than rely on this single path to set up everything, instead have a `ControlsState` enum that
`HTMLMediaElement` can internally reason about to decide whether or not it needs to inject the JS
that creates `createControls` and then call it. This also has the added benefit of having those two
steps now be in the same place, whereas before they were split across two unrelated functions.

* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::HTMLMediaElement):
(WebCore::HTMLMediaElement::updateCaptionContainer):
(WebCore::HTMLMediaElement::virtualHasPendingActivity const):
(WebCore::HTMLMediaElement::configureTextTrackDisplay):
(WebCore::HTMLMediaElement::updateTextTrackDisplay):
(WebCore::HTMLMediaElement::updateTextTrackRepresentationImageIfNeeded):
(WebCore::HTMLMediaElement::configureMediaControls):
(WebCore::HTMLMediaElement::ensureMediaControls): Added.
(WebCore::HTMLMediaElement::getCurrentMediaControlsStatus):
(WebCore::HTMLMediaElement::ensureMediaControlsShadowRoot): Deleted.
(WebCore::HTMLMediaElement::ensureMediaControlsInjectedScript): Deleted.
(WebCore::HTMLMediaElement::didAddUserAgentShadowRoot): Deleted.

* LayoutTests/media/audio-controls-adoptNode.html: Added.
* LayoutTests/media/audio-controls-adoptNode-expected.txt: Added.
* LayoutTests/media/video-controls-adoptNode.html: Added.
* LayoutTests/media/video-controls-adoptNode-expected.txt: Added.

Canonical link: https://commits.webkit.org/253225@main
